src/basics.ml

```diff
@@ -240,6 +240,16 @@ let distinct_vars : term array -> bool =
     | _ -> raise Not_unique_var
   in try Array.iter check ts; true with Not_unique_var -> false
 
+(** [is_pattern t] tells whether term [t] is a valid pattern. *)
+let rec is_pattern : term -> bool = fun t ->
+  match get_args (unfold t) with
+  | (Patt(_), []  ) -> true (* Pattern variables are not applied *)
+  | (Patt(_), args) -> distinct_vars (Array.of_list args)
+  | (Abst(_,b), []) ->
+      let _, b = Bindlib.unbind b in
+      is_pattern b
+  | (_      , args) -> List.for_all is_pattern args
```

Comment:
don't you need to check that you have a function symbol at the head? can your trees handle a pattern like (\x, x &F)?

Reply:
Some of the checks are already done in scoping. Maybe that would be a better place to extend the tests.